tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners/Program.cs

```diff
@@ -1,54 +1,118 @@
 using Azure.Sdk.Tools.CodeOwnersParser;
 
-namespace Azure.Sdk.Tools.RetrieveCodeOwners
+namespace Azure.Sdk.Tools.RetrieveCodeOwners;
+
+/// <summary>
+/// See Program.Main comment.
+/// </summary>
+public static class Program
 {
     /// <summary>
-    /// See Program.Main comment.
+    /// Given targetPath and CODEOWNERS file path or https url codeownersFilePathOrUrl,
+    /// prints out to stdout owners of the targetPath as determined by the CODEOWNERS data.
     /// </summary>
-    public static class Program
+    /// <param name="targetPath">The path whose owners are to be determined. Can be a glob path.</param>
+    /// <param name="codeownersFilePathOrUrl">The https url or path to the CODEOWNERS file.</param>
+    /// <param name="excludeNonUserAliases">Whether owners that aren't users should be excluded from the
+    /// returned owners.</param>
+    /// <param name="targetDir">
+    /// The directory to search for file paths in case targetPath is a glob path Unused otherwise.
+    /// </param>
+    /// <param name="useRegexMatcher">
+    /// Whether to use the new matcher for CODEOWNERS file paths, that supports matching
+    /// entries with * and **, instead of silently ignoring them.
+    /// </param>
+    /// <returns>
+    /// On STDOUT: The JSON representation of the matched CodeownersEntry.
+    /// "new CodeownersEntry()" if no path in the CODEOWNERS data matches.
+    /// <br/><br/>
+    /// From the Main method: exit code. 0 if successful, 1 if error.
+    /// </returns>
+    public static int Main(
+        string targetPath,
+        string codeownersFilePathOrUrl,
+        bool excludeNonUserAliases = false,
+        string? targetDir = null,
+        bool useRegexMatcher = false)
     {
-        /// <summary>
-        /// Given targetPath and CODEOWNERS file path or https url codeownersFilePathOrUrl,
-        /// prints out to stdout owners of the targetPath as determined by the CODEOWNERS data.
-        /// </summary>
-        /// <param name="targetPath">The path whose owners are to be determined.</param>
-        /// <param name="codeownersFilePathOrUrl">The https url or path to the CODEOWNERS file.</param>
-        /// <param name="excludeNonUserAliases">Whether owners that aren't users should be excluded from the
-        /// returned owners.</param>
-        /// <returns>
-        /// On STDOUT: The JSON representation of the matched CodeownersEntry.
-        /// "new CodeownersEntry()" if no path in the CODEOWNERS data matches.
-        /// <br/><br/>
-        /// From the Main method: exit code. 0 if successful, 1 if error.
-        /// </returns>
-        public static int Main(
-            string targetPath,
-            string codeownersFilePathOrUrl,
-            bool excludeNonUserAliases = false)
+        targetPath = targetPath.Trim();
+        targetDir = targetDir?.Trim();
+        codeownersFilePathOrUrl = codeownersFilePathOrUrl.Trim();
+
+        Debug.Assert(!string.IsNullOrWhiteSpace(targetPath));
```

Comment:
Why not make these actual runtime checks?

Reply:
@weshaggard good point; I am going to make it throw always, but I will catch it, to print out to STDERR and return exit code of 1. 

This is because currently, only [the get-codeowners.ps1, uses this tool](https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/get-codeowners.ps1#L41-L57), like so:

``` ps1
  if($includeNonUserAliases) {
    $codeOwnersString = & $command --target-directory $targetDirectory --code-owner-file-path $codeOwnerFileLocation 2>&1
  }
  else {
    $codeOwnersString = & $command --target-directory $targetDirectory --code-owner-file-path $codeOwnerFileLocation --filter-out-non-user-aliases 2>&1
  }
  # Failed at the command of fetching code owners.
  if ($LASTEXITCODE -ne 0) {
    Write-Host $codeOwnersString
    return ,@()
  }
  
  $codeOwnersJson = $codeOwnersString | ConvertFrom-Json
  if (!$codeOwnersJson) {
    Write-Host "No code owners returned from the path: $targetDirectory"
    return ,@()
  }
```

As you can see, it doesn't throw upon failure, only prints out the message and returns empty array. Hence, at least in this PR, I want to stay compatible with that usage.